File: backend/app/routers/alerts.py
from datetime import date, timedelta

from fastapi import APIRouter, Depends, Query
from sqlalchemy import select
from sqlalchemy.orm import Session, selectinload

from app.database import get_db
from app.models import Part
from app.schemas import PartOut

router = APIRouter(prefix="/alerts", tags=["alerts"])


@router.get("/warranty", response_model=list[PartOut])
def warranty_alerts(
    db: Session = Depends(get_db),
    days: int = Query(30, ge=1, le=365),
):
    today = date.today()
    cutoff = today + timedelta(days=days)
    parts = (
        db.execute(
            select(Part)
            .options(selectinload(Part.pc))
            .where(Part.warranty_expiry.is_not(None))
            .where(Part.warranty_expiry >= today)
            .where(Part.warranty_expiry <= cutoff)
            .order_by(Part.warranty_expiry.asc())
        )
        .scalars()
        .all()
    )
    out = []
    for part in parts:
        item = PartOut.model_validate(part)
        item.pc_name = part.pc.name if part.pc else None
        out.append(item)
    return out

File: backend/routes/alerts.js
import { Router } from 'express';
import { Part, PC } from '../config/models.js';

const router = Router();

router.get('/warranty', async (req, res, next) => {
  try {
    const days = Math.min(parseInt(req.query.days) || 30, 365);
    const today = new Date();
    const cutoff = new Date(today.getTime() + days * 86400000);
    let parts = await Part.find({}).lean();
    parts = parts.filter(p => p.warranty_expiry && new Date(p.warranty_expiry) >= today && new Date(p.warranty_expiry) <= cutoff);
    parts.sort((a, b) => new Date(a.warranty_expiry) - new Date(b.warranty_expiry));
    const result = await Promise.all(parts.map(async p => {
      let pc_name = null;
      if (p.pc_id) { const pc = await PC.findById(p.pc_id).lean(); if (pc) pc_name = pc.name; }
      return { ...p, id: p._id, pc_name };
    }));
    res.json(result);
  } catch (err) { next(err); }
});

export default router;
